File: web/src/lib/taskTiming.test.ts
import { describe, it, expect } from 'vitest'
import {
  pickupLatencyMs,
  inProgressDurationMs,
  isOverThreshold,
  formatCoarseDuration,
} from './taskTiming'
import type { Task } from './types'

// A fixed wall-clock anchor so spans are deterministic regardless of run time.
const T0 = Date.parse('2026-05-10T12:00:00.000Z')
const at = (ms: number) => new Date(T0 + ms).toISOString()
const MIN = 60_000
const HOUR = 60 * MIN
const DAY = 24 * HOUR

// Minimal task factory — only the fields the timing helpers read.
function task(partial: Partial<Task>): Task {
  return { id: 'task-1', description: 'do the thing', status: 'pending', ...partial }
}

describe('pickupLatencyMs', () => {
  it('returns null when never claimed', () => {
    expect(pickupLatencyMs(task({ createdAt: at(0) }))).toBeNull()
  })
  it('returns null when creation time unknown', () => {
    expect(pickupLatencyMs(task({ startedAt: at(MIN) }))).toBeNull()
  })
  it('returns startedAt - createdAt', () => {
    expect(pickupLatencyMs(task({ createdAt: at(0), startedAt: at(3 * MIN) }))).toBe(3 * MIN)
  })
  it('clamps clock skew to zero', () => {
    expect(pickupLatencyMs(task({ createdAt: at(5 * MIN), startedAt: at(MIN) }))).toBe(0)
  })
})

describe('inProgressDurationMs', () => {
  const now = T0 + HOUR
  it('returns null when never started', () => {
    expect(inProgressDurationMs(task({ createdAt: at(0) }), now)).toBeNull()
  })
  it('uses now while in flight', () => {
    expect(inProgressDurationMs(task({ startedAt: at(10 * MIN) }), now)).toBe(50 * MIN)
  })
  it('uses completedAt when finished', () => {
    expect(
      inProgressDurationMs(task({ startedAt: at(10 * MIN), completedAt: at(25 * MIN) }), now),
    ).toBe(15 * MIN)
  })
  it('clamps clock skew to zero', () => {
    expect(
      inProgressDurationMs(task({ startedAt: at(30 * MIN), completedAt: at(10 * MIN) }), now),
    ).toBe(0)
  })
})

describe('isOverThreshold', () => {
  const now = T0 + HOUR
  it('flags an in-flight task past the threshold', () => {
    expect(isOverThreshold(task({ status: 'active', startedAt: at(0) }), 30 * MIN, now)).toBe(true)
  })
  it('does not flag a task under the threshold', () => {
    expect(isOverThreshold(task({ status: 'active', startedAt: at(0) }), 90 * MIN, now)).toBe(false)
  })
  it('never flags a terminal task', () => {
    const t = task({ status: 'succeeded', startedAt: at(0), completedAt: at(90 * MIN) })
    expect(isOverThreshold(t, MIN, now)).toBe(false)
  })
  it('never flags an unstarted task', () => {
    expect(isOverThreshold(task({ status: 'pending', createdAt: at(0) }), MIN, now)).toBe(false)
  })
  it('treats a non-positive threshold as disabled', () => {
    expect(isOverThreshold(task({ status: 'active', startedAt: at(0) }), 0, now)).toBe(false)
  })
})

describe('formatCoarseDuration', () => {
  it('passes null through', () => {
    expect(formatCoarseDuration(null)).toBeNull()
  })
  it('shows <1m under a minute', () => {
    expect(formatCoarseDuration(30_000)).toBe('<1m')
  })
  it('shows minutes under an hour', () => {
    expect(formatCoarseDuration(45 * MIN)).toBe('45m')
  })
  it('shows hours and minutes', () => {
    expect(formatCoarseDuration(2 * HOUR + 15 * MIN)).toBe('2h 15m')
  })
  it('drops minutes on a whole hour', () => {
    expect(formatCoarseDuration(3 * HOUR)).toBe('3h')
  })
  it('shows days and hours past a day', () => {
    expect(formatCoarseDuration(DAY + 4 * HOUR)).toBe('1d 4h')
  })
  it('drops hours on a whole day', () => {
    expect(formatCoarseDuration(2 * DAY)).toBe('2d')
  })
})

File: internal/services/task/domain/task_timing_test.go
package domain

import (
	"testing"
	"time"
)

// tp returns a pointer to a time offset from fixedTime (declared in
// task_aggregate_test.go) by the given duration — keeps the timing cases terse.
func tp(d time.Duration) *time.Time {
	t := fixedTime.Add(d)
	return &t
}

func TestPickupLatency(t *testing.T) {
	t.Run("nil when never claimed", func(t *testing.T) {
		task := Task{CreatedAt: tp(0)}
		if got := task.PickupLatency(); got != nil {
			t.Fatalf("PickupLatency=%v want nil", got)
		}
	})
	t.Run("nil when creation time unknown", func(t *testing.T) {
		task := Task{StartedAt: tp(time.Minute)}
		if got := task.PickupLatency(); got != nil {
			t.Fatalf("PickupLatency=%v want nil", got)
		}
	})
	t.Run("started minus created", func(t *testing.T) {
		task := Task{CreatedAt: tp(0), StartedAt: tp(3 * time.Minute)}
		got := task.PickupLatency()
		if got == nil || *got != 3*time.Minute {
			t.Fatalf("PickupLatency=%v want 3m", got)
		}
	})
	t.Run("clamps clock skew to zero", func(t *testing.T) {
		task := Task{CreatedAt: tp(5 * time.Minute), StartedAt: tp(time.Minute)}
		got := task.PickupLatency()
		if got == nil || *got != 0 {
			t.Fatalf("PickupLatency=%v want 0", got)
		}
	})
}

func TestInProgressDuration(t *testing.T) {
	now := fixedTime.Add(time.Hour)
	t.Run("nil when never started", func(t *testing.T) {
		task := Task{CreatedAt: tp(0)}
		if got := task.InProgressDuration(now); got != nil {
			t.Fatalf("InProgressDuration=%v want nil", got)
		}
	})
	t.Run("uses now while in flight", func(t *testing.T) {
		task := Task{StartedAt: tp(10 * time.Minute)} // started, not completed
		got := task.InProgressDuration(now)
		want := 50 * time.Minute // now is +60m, started +10m
		if got == nil || *got != want {
			t.Fatalf("InProgressDuration=%v want %v", got, want)
		}
	})
	t.Run("uses completedAt when finished", func(t *testing.T) {
		task := Task{StartedAt: tp(10 * time.Minute), CompletedAt: tp(25 * time.Minute)}
		got := task.InProgressDuration(now)
		if got == nil || *got != 15*time.Minute {
			t.Fatalf("InProgressDuration=%v want 15m", got)
		}
	})
	t.Run("clamps clock skew to zero", func(t *testing.T) {
		task := Task{StartedAt: tp(30 * time.Minute), CompletedAt: tp(10 * time.Minute)}
		got := task.InProgressDuration(now)
		if got == nil || *got != 0 {
			t.Fatalf("InProgressDuration=%v want 0", got)
		}
	})
}

func TestIsOverThreshold(t *testing.T) {
	now := fixedTime.Add(time.Hour)
	t.Run("true when in-flight task exceeds threshold", func(t *testing.T) {
		task := Task{Status: TaskStatusActive, StartedAt: tp(0)} // in progress 60m
		if !task.IsOverThreshold(30*time.Minute, now) {
			t.Fatal("expected breach for 60m > 30m threshold")
		}
	})
	t.Run("false when under threshold", func(t *testing.T) {
		task := Task{Status: TaskStatusActive, StartedAt: tp(0)} // 60m
		if task.IsOverThreshold(90*time.Minute, now) {
			t.Fatal("expected no breach for 60m < 90m threshold")
		}
	})
	t.Run("terminal tasks never breach", func(t *testing.T) {
		task := Task{Status: TaskStatusSucceeded, StartedAt: tp(0), CompletedAt: tp(90 * time.Minute)}
		if task.IsOverThreshold(time.Minute, now) {
			t.Fatal("terminal task should never breach")
		}
	})
	t.Run("never-started task never breaches", func(t *testing.T) {
		task := Task{Status: TaskStatusPending, CreatedAt: tp(0)}
		if task.IsOverThreshold(time.Minute, now) {
			t.Fatal("unstarted task should never breach")
		}
	})
	t.Run("non-positive threshold disables the check", func(t *testing.T) {
		task := Task{Status: TaskStatusActive, StartedAt: tp(0)}
		if task.IsOverThreshold(0, now) {
			t.Fatal("threshold of 0 should disable the check")
		}
	})
}
